runtime/sema/check_interface_declaration.go

```diff
@@ -366,3 +386,137 @@ func (checker *Checker) declareInterfaceMembers(declaration *ast.InterfaceDeclar
 		checker.declareAttachmentMembersAndValue(nestedAttachmentDeclaration, ContainerKindInterface)
 	}
 }
+
+func (checker *Checker) declareEntitlementType(declaration *ast.EntitlementDeclaration) *EntitlementType {
+	identifier := declaration.Identifier
+
+	entitlementType := NewEntitlementType(checker.memoryGauge, checker.Location, identifier.Identifier)
+
+	variable, err := checker.typeActivations.declareType(typeDeclaration{
+		identifier:               identifier,
+		ty:                       entitlementType,
+		declarationKind:          declaration.DeclarationKind(),
+		access:                   checker.accessFromAstAccess(declaration.Access),
+		docString:                declaration.DocString,
+		allowOuterScopeShadowing: false,
+	})
+
+	checker.report(err)
+	if checker.PositionInfo != nil && variable != nil {
+		checker.recordVariableDeclarationOccurrence(
+			identifier.Identifier,
+			variable,
+		)
+	}
+
+	checker.Elaboration.SetEntitlementDeclarationType(declaration, entitlementType)
+	checker.Elaboration.SetEntitlementTypeDeclaration(entitlementType, declaration)
+
+	return entitlementType
+}
+
+func (checker *Checker) VisitEntitlementDeclaration(declaration *ast.EntitlementDeclaration) (_ struct{}) {
+
+	entitlementType := checker.Elaboration.EntitlementDeclarationType(declaration)
+	if entitlementType == nil {
+		panic(errors.NewUnreachableError())
+	}
+
+	checker.checkDeclarationAccessModifier(
+		checker.accessFromAstAccess(declaration.Access),
+		declaration.DeclarationKind(),
+		entitlementType,
+		nil,
+		declaration.StartPos,
+		true,
+	)
+
+	return
+}
+
+func (checker *Checker) declareEntitlementMappingType(declaration *ast.EntitlementMappingDeclaration) *EntitlementMapType {
+	identifier := declaration.Identifier
+
+	entitlementMapType := NewEntitlementMapType(checker.memoryGauge, checker.Location, identifier.Identifier)
+
+	variable, err := checker.typeActivations.declareType(typeDeclaration{
+		identifier:               identifier,
+		ty:                       entitlementMapType,
+		declarationKind:          declaration.DeclarationKind(),
+		access:                   checker.accessFromAstAccess(declaration.Access),
+		docString:                declaration.DocString,
+		allowOuterScopeShadowing: false,
+	})
+
+	checker.report(err)
+	if checker.PositionInfo != nil && variable != nil {
+		checker.recordVariableDeclarationOccurrence(
+			identifier.Identifier,
+			variable,
+		)
+	}
+
+	checker.Elaboration.SetEntitlementMapDeclarationType(declaration, entitlementMapType)
+	checker.Elaboration.SetEntitlementMapTypeDeclaration(entitlementMapType, declaration)
+
+	return entitlementMapType
+}
+
+func (checker *Checker) declareEntitlementMappingElements(declaration *ast.EntitlementMappingDeclaration) {
+
+	entitlementMapType := checker.Elaboration.EntitlementMapDeclarationType(declaration)
+	if entitlementMapType == nil {
```

Comment:
Same as above, a one line comment explaining why this can never fail would be helpful